src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1050,7 +1106,8 @@ DistributedExecutionRequiresRollback(DistributedExecution *execution)
 
 	if (ReadOnlyTask(task->taskType))
 	{
-		return SelectOpensTransactionBlock && IsTransactionBlock();
+		return !execution->isOutsideTransaction && SelectOpensTransactionBlock &&
+			   IsTransactionBlock();
```

Comment:
Maybe add an `if (execution->isOutsideTransaction) return false;` somewhere above to make it more explicit and robust to other types of tasks.